ext/app/electron/DocRegistry.ts

```diff
@@ -0,0 +1,74 @@
+import * as path from "path";
+import { HomeDBManager } from "app/gen-server/lib/homedb/HomeDBManager";
+import { fileExists } from "./utils";
+
+export class DocRegistry {
+
+  private idToPathMap: Map<string, string>;
+  private pathToIdMap: Map<string, string>;
+  private db: HomeDBManager;
+
+  private constructor(dbManager: HomeDBManager) {
+    this.db = dbManager;
+  }
+
+  public static async create(dbManager: HomeDBManager) {
+    // Allocate space.
+    const dr = new DocRegistry(dbManager);
+    dr.idToPathMap = new Map<string, string>;
+    dr.pathToIdMap = new Map<string, string>;
+
+    // Go over all documents we know about.
+    for (const doc of await dr.db.getAllDocs()) {
+      // All documents are supposed to have externalId set.
+      const docPath = doc.options?.externalId;
+      if (docPath && fileExists(docPath)) {
+        // Cache the two-way mapping docID <-> path.
+        dr.idToPathMap.set(doc.id, docPath);
+        dr.pathToIdMap.set(docPath, doc.id);
+      } else {
+        // Remove this document - it should not appear in a DB for Grist Desktop.
+        await dr.db.deleteDocument({
+          userId: (await dr.getDefaultUser()).id,
+          urlId: doc.id
+        });
+      }
+    }
+    return dr;
+  }
```

Comment:
I'm concerned about this class being a cache over the top of home DB - as it could introduce consistency issues. Is there a strong reason we need it?

Reply:
I responded in the other comment you left about cache coherence. Marking this as resolved as it's duplicate.